src/algos/dtmscc.jl

```diff
@@ -0,0 +1,286 @@
+
+using Optim
+
+using Interpolations
+using splines # temporary
+
+#####
+# helper functions/types (to be moved/removed)
+#####
+
+mlinspace(a,b,orders) = gridmake([linspace(a[i],b[i],orders[i]) for i=1:length(orders)]...)
+
+
+type MarkovChain
+    P::Array{Float64, 2}
+    Q::Array{Float64, 2}
+end
+
+type ApproximationSpace
+  a:: Array{Float64, 1}
+  b:: Array{Float64, 1}
+  orders:: Array{Int64, 1}
+end
+
+
+####
+# residuals
+####
+
+function residuals(model::DTMSCCModel, calibration::ModelCalibration)
+    m = calibration[:markov_states]
+    s = calibration[:states]
+    x = calibration[:controls]
+    p = calibration[:parameters]
+    res = Dict()
+    res[:transition] = (s-evaluate(model.functions.transition,m,s,x,m,p))
+    res[:arbitrage] = (evaluate(model.functions.arbitrage,m,s,x,m,s,x,p))
+    return res
+end
+
+residuals(model::DTMSCCModel) = residuals(model, model.calibration)
+
+####
+# construction of initial guess
+####
+
+function constant_guess(model::DTMSCCModel)
+    x = model.calibration[:controls]
+    function fun(i,s)
+        if ndims(s) == 1
+            return x
+        else
+            N = size(s,1)
+            return repmat(x',N,1)
+        end
+    end
+    return fun
+end
+
+####
+# evaluate a given policy rule (a function)
+####
+
+function eval_policy(model::DTMSCCModel, policy, options=Dict())
+
+    felicity = model.functions.felicity
+    transition = model.functions.transition
+
+    discount = model.calibration.flat[:beta]
+
+    # get approximation space
+    ap = model.options[:approximation_space]
+    a = ap[:a]
+    b = ap[:b]
+    orders = ap[:orders]
+    approx = ApproximationSpace(a,b,orders)
+
+    grid = mlinspace(a,b,orders)
+    d = size(grid,2)
+
+    # get markov chain
+    mcopts = model.options[:discrete_transition][:MarkovChain]
+    P = cat(1, [e' for e  in mcopts[1]]...)
+    Q = cat(1, [e' for e  in mcopts[2]]...)
+    markov_chain = MarkovChain(P,Q)
+    nodes = markov_chain.P
+
+    symbols = model.symbolic.symbols
+    calibration = model.calibration
+
+    N = size(grid,1)
+    n_mc = size(markov_chain.P,1)
+    n_s = size(grid,2)
+    n_x = length(symbols[:controls])
+    p = calibration
+    controls = zeros(n_mc,N,n_x)
+    value = zeros(n_mc,N)
+    p = calibration[:parameters]
+    for i_mc=1:n_mc
+        for n=1:N
+            s = slice(grid,n,:)
+            m = slice(nodes,i_mc,:)
+            x = policy(i_mc,s)
+            controls[i_mc, n,:] = x
+            value[i_mc, n:n] *= evaluate(felicity, m, s, x, p)/(1-discount)
+        end
+    end
+
+    print(value)
+    it = 0
+    maxit = get(options, :maxit, 500)
+    tol =   get(options, :tol, 1e-6)
+    err = 10
+
+    dims = vcat(approx.orders)
+    knots = [collect(linspace(approx.a[i], approx.b[i], approx.orders[i])) for i=1:d]
+    knots = tuple(knots...)
+
+    create_interpolant(vv) = [interpolate(knots, copy(reshape(slice(vv,i_mc,:),dims...)), Gridded(Linear()))  for i_mc in 1:n_mc ]
```

Comment:
Here you are using `Gridded(Linear())` because the knots are all originally `linespace` objects, it would be more efficient to use 

```
knots = [linspace(approx.a[i], approx.b[i], approx.orders[i]) for i=1:d]
scale(interpolate(copy(reshape(slice(vv,i_mc,:),dims...)), BSpline(Linear()), OnGrid()), knots...)
```

Not only will this be more efficient, but scaling works on any interpolation types, so you can do things like 

```
scale(interpolate(copy(reshape(slice(vv,i_mc,:),dims...)), BSpline(Cubic(Flat())), OnGrid()), knots...)
```